Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+ 
+Moreover there is an 'obvious' way of extending the CIL file to include the additional data we need. A CLI file has a well-defined header structure, and that header already has a field that can point of to 'additional information'. This is used today in NGEN images. We would use this same technique to allow the existing CLI format to include a new 'Native Header' that would then point at any additional information needed to support fast, direct execution. 
+
+The most important parts of this extra information include
+
+1. Native code for the methods (as well as a way of referencing things outside the module)
+2. Garbage Collection (GC) information for each method that allows you to know what values in registers and on the stack are pointers to the GC heap wherever a GC is allowed. 
+3. Exception handling (EH) tables that allow an exception handler to be found when an exception is thrown. 
+4. A table that allows the GC and EH to be found given just the current instruction pointer (IP) within the code. (IP map).
+5. A table that links the information in the metadata to the corresponding native structure. 
+
+That is, we need something to link the world of metadata to the world of native. We can't eliminate meta-data completely because we want to support existing functionality. In particular we need to be able to support having other CLI images refer to types, methods and fields in this image. They will do so by referencing the information in the metadata, but once they find the target in the metadata, we will need to find the actual native code or type information corresponding to that meta-data entry. This is the purpose of the additional table. Effectively, this table is the 'export' mechanism for managed references.
+
+Some of this information can be omitted or stored in more efficient form, e.g.:
+
+- The garbage collection information can be omitted for environments with conservative garbage collection, such as IL2CPP.
+- The full metadata information is not strictly required for 'private' methods or types so it is possible to strip it from the CLI image.
+- The metadata can be store in more efficient form, such as the .NET Native metadata format.
+- The platform native executable format (ELF, Mach-O) can be used as envelope instead of PE to take advantage of platform OS loader.
+
+
+## Definition of Version Compatibility for Native Code
+
+Even for IL or unmanaged native code, there are limits to what compatible changes can be made. For example, deleting a public method is sure to be an incompatible change for any extern code using that method.
+ 
+Since CIL already has a set of [compatibility rules](https://github.com/dotnet/corefx/wiki/Breaking-Changes), ideally the native format would have the same set of compatibility rules as CIL. Unfortunately, that is difficult to do efficiently in all cases. In those cases we have multiple choices
```

Comment:
That link doesn't exist anymore, it should be https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-changes.md